private/buf/buflsp/file.go

```diff
@@ -285,9 +289,28 @@ func (f *file) Refresh(ctx context.Context) {
 	f.FindModule(ctx)
 
 	progress.Report(ctx, "Running Checks", 4.0/6)
-	f.BuildImages(ctx)
-	f.RunLints(ctx)
-	f.RunBreaking(ctx)
+	// Since checks are a more expensive operation, we do not want to run a check on every
+	// Refresh call. Instead, we can stagger the checks and only run them periodically by
+	// spinning them off into a go routine. Then we attempt to lock using the top-level LSP
+	// lock. It is safe to use because if another LSP call is made, we allow checks to finish
+	// before resolving a subsequent LSP request.
+	go func() {
+		// Call TryLock, if unnsuccessful, then another thread holds the lock, so we provide a
+		// debug log and move on.
+		if !f.lsp.lock.TryLock() {
```

Comment:
This looks like the final run of lint and breaking could be mismatched to the output. I think we'd want to ensure if a change occurred after the job was run but no change happens after its finished it re-runs to reflect the latest. This could also be done by cancelling the previous work and always starting at the latest.

Reply:
The way the stagger works is that we do not want to hammer checks when many `Refresh` calls are made in a flurry. I just fixed the ordering, so we should be converging on the final state when the checks are run. The top-level locking would also mean you wouldn't need to cancel the previous work because no additional LSP requests are being handled while the checks are running.

Reply:
The number of running jobs should be limited but the final annotations should always catch up to the latest otherwise I think the output could be stale e.g.:

```mermaid
sequenceDiagram
    Client->>Refresh: Change triggered job
    activate Refresh
    Client--xRefresh: Change before job is completed is lost
    Refresh->>Client: Stale annotation
    deactivate Refresh
```
Could mean that the user just has to write a comment or something to trigger the job again? The lock could also mean the job is always skipped if the LSP is busy?

Reply:
It's the top-level LSP lock, so if a job is running, then the next `Refresh` request blocks until the lock is freed up again, so this should not be the case...